src/collection-view.js

```diff
@@ -125,11 +125,90 @@ const CollectionView = Backbone.View.extend({
     }
   },
 
-  // get the child view by model it holds, and remove it
-  _onCollectionRemove(model) {
-    const view = this.children.findByModel(model);
-    this.removeChildView(view);
-    this._checkEmpty();
+  // Handle collection update model removals
+  _onCollectionUpdate(collection, options) {
+    const changes = options.changes;
+    if (changes) {
+      const removed = changes.removed;
+      if (_.isArray(removed)) {
+        this._removeChildViews(removed);
+      }
+    }
+  },
+
+  // Remove the child views and destroy them.
+  // This function also updates the indices of later views
+  // in the collection in order to keep the children in sync with the collection.
+  // "modelsOrViews" can be an array of mixed models or views that
+  // will be removed and destroyed from the CollectionView
+  _removeChildViews(modelsOrViews, {checkEmpty} = {}) {
+    if (!modelsOrViews.length) {
+      return modelsOrViews;
+    }
+
+    const shouldCheckEmpty = checkEmpty !== false;
+
+    // Used to determine where to update the remaining
+    // sibling view indices after these views are removed.
+    const removedViews = this._getRemovedViews(modelsOrViews);
+
+    if (!removedViews.length) {
+      return removedViews;
+    }
+
+    this.children._updateLength();
+    // decrement the index of views after this one
+    this._updateIndices(removedViews[0], false);
+
+    if (shouldCheckEmpty) {
+      this._checkEmpty();
+    }
+
+    return removedViews;
+  },
+
+  // "modelsOrViews" can be an array of mixed models or views that
+  // will be removed and destroyed from the CollectionView
+  // Returns an array of views where the first view has the index that
+  // tells CollectionView#_updateIndices where to start syncing remaining
+  // view indices.
+  _getRemovedViews(modelsOrViews) {
+    return _.reduce(modelsOrViews, (removingViews, modelOrView) => {
+      let view = modelOrView;
+
+      if (modelOrView instanceof Backbone.Model) {
+        view = this.children.findByModel(modelOrView);
+      }
+
+      if (!view || view._isDestroyed) {
+        return removingViews;
+      }
+
+      if (!removingViews.length || removingViews[0]._index > view._index) {
+        removingViews.unshift(view);
+      } else {
+        removingViews.push(view);
+      }
+
+      this._destroyChildView(view);
+
+      return removingViews;
+    }, []);
+  },
+
+  _destroyChildView(view) {
+    this.triggerMethod('before:remove:child', this, view);
+
+    this.children._remove(view);
+    if (view.destroy) {
+      view.destroy();
+    } else {
+      destroyBackboneView(view);
+    }
+
+    view._parent = void 0;
```

Comment:
while this might be a micro-optimization we should change it stylistically across the code rather than introduce the change in a single location for this PR, so that future contributors know which to do.  Granted they are slightly different on the prototype.. however I doubt we rely on that difference in this lib.